theano/sandbox/cuda/dnn.py

```diff
@@ -2270,6 +2270,375 @@ def method(self):
+        assert x.ndim == 4
+        assert scale.ndim == 4
+        assert bias.ndim == 4
+        assert estimated_mean.ndim == 4
+        assert estimated_variance.ndim == 4
+        return Apply(self, [x, scale, bias, estimated_mean, estimated_variance],
+                     [x.type()])
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        # super call to prepare common configuration
+        result = super(GpuDnnBatchNormInference, self).c_code(node, name, inputs, outputs, sub)
+
+        # give sensible names to inputs and outputs
+        inp, scale, bias, est_mean, est_var = inputs
+        outp, = outputs
+
+        # set input tensor descriptors from input tensors
+        result += c_set_tensor4d(inp, 'bn_input_' + name, 'err' + name, sub['fail'])
+        result += c_set_tensor4d(scale, 'bn_params_' + name, 'err' + name, sub['fail'])
+
+        # build and prepare the output variable
+        result += """
+if (CudaNdarray_prep_output(&%(outp)s, 4, CudaNdarray_HOST_DIMS(%(inp)s)) != 0)
+{
+    %(fail)s
+}
+""" % dict(outp=outp, inp=inp, fail=sub['fail'])
+
+        # set output tensor descriptor from output tensor
+        result += c_set_tensor4d(outp, 'bn_output_' + name, 'err' + name, sub['fail'])
+
+        # call cuDNN function
+        result += """
+{
+const float alpha = 1.;
+const float beta = 0.;
+err%(name)s = cudnnBatchNormalizationForwardInference(
+  _handle,
+  mode%(name)s,
+  (void*) &alpha,
+  (void*) &beta,
+  bn_input_%(name)s,
+  CudaNdarray_DEV_DATA(%(inp)s),
+  bn_output_%(name)s,
+  CudaNdarray_DEV_DATA(%(outp)s),
+  bn_params_%(name)s,
+  CudaNdarray_DEV_DATA(%(scale)s),
+  CudaNdarray_DEV_DATA(%(bias)s),
+  CudaNdarray_DEV_DATA(%(est_mean)s),
+  CudaNdarray_DEV_DATA(%(est_var)s),
+  epsilon%(name)s
+);
+}
+""" % dict(name=name, inp=inp, scale=scale, bias=bias, est_mean=est_mean,
+           est_var=est_var, outp=outp)
+
+        return result
+
+    def grad(self, inputs, grads):
+        x, scale, bias, est_mean, est_var = inputs
+        dy = grads[0]
+
+        # add necessary broadcasts
+        if self.mode == 'per-activation':
+            axes = (0,)
+        elif self.mode == 'spatial':
+            axes = (0, 2, 3)
+        scale, bias, est_mean, est_var = (theano.tensor.addbroadcast(t, *axes)
+                                          for t in (scale, bias, est_mean, est_var))
+
+        # define helper expressions
+        est_var_eps = est_var + self.epsilon
+        est_std = theano.tensor.sqrt(est_var_eps)
+        two = theano.tensor.constant(2.)
+
+        # define and return gradients
+        dx = dy * (scale / est_std)
+        dscale = (dy * (x - est_mean)).sum(axes, keepdims=True) / est_std
+        dbias = dy.sum(axes, keepdims=True)
+        dmean = -dy.sum(axes, keepdims=True) * (scale / est_std)
+        dvar = -(dy * (x - est_mean)).sum(axes, keepdims=True) * (scale / (two * est_var_eps * est_std))
```

Comment:
I don't get why for those case they can't take the training implementation
of cudnn bn?

On Wed, Jun 15, 2016 at 1:02 PM, ballasn <notifications@github.com> wrote:

> In theano/sandbox/cuda/dnn.py
> <https://github.com/Theano/Theano/pull/4582#discussion_r67202361>:
>
> > +        elif self.mode == 'spatial':
> > +            axes = (0, 2, 3)
> > +        scale, bias, est_mean, est_var = (theano.tensor.addbroadcast(t, *axes)
> > +                                          for t in (scale, bias, est_mean, est_var))
> > +
> > +        # define helper expressions
> > +        est_var_eps = est_var + self.epsilon
> > +        est_std = theano.tensor.sqrt(est_var_eps)
> > +        two = theano.tensor.constant(2.)
> > +
> > +        # define and return gradients
> > +        dx = dy * (scale / est_std)
> > +        dscale = (dy * (x - est_mean)).sum(axes, keepdims=True) / est_std
> > +        dbias = dy.sum(axes, keepdims=True)
> > +        dmean = -dy.sum(axes, keepdims=True) * (scale / est_std)
> > +        dvar = -(dy * (x - est_mean)).sum(axes, keepdims=True) * (scale / (two * est_var_eps * est_std))
>
> I see, ok then for the gradient in the inference op.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/4582/files/c4a9ee20114bb4c0adc9000c8d424f2cfc44fa3a..d63f5ee32aa10369eba3dbd4c266b7b833b70d26#r67202361>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AALC-wMQUyV13mjiquerFpjb1-16HuD7ks5qMDAngaJpZM4Ituzu>
> .
>